servicetalk-concurrent-api-internal/src/test/java/io/servicetalk/concurrent/api/single/AbstractPublishAndSubscribeOnTest.java

```diff
@@ -81,4 +85,54 @@
         verifyCapturedThreads(capturedThreads);
         return capturedThreads;
     }
+
+    private static void recordThread(AtomicReferenceArray<Thread> threads, final int index) {
+        Thread was = threads.getAndUpdate(index, AbstractPublishAndSubscribeOnTest::updateThread);
+        assertThat("Thread already recorded at index: " + index, was, nullValue());
```

Comment:
You have duplicated validation for the `was` value. One is here with the message, another one is inside `updateThread` that will fail without a context of the index. Consider removing `assertThat(current, nullValue())` there.